File: mac-app/QuartzTeachers/Models/Course.swift
import Foundation
import Observation

/// One course (or club) folder inside the working folder's `courses/`
/// directory, together with its loaded configuration.
@Observable
class Course: Identifiable {

    // MARK: - Stored properties

    /// The course code, e.g. "ICS3U" — also the folder name.
    let code: String

    /// The course folder, e.g. `<workspace>/courses/ICS3U`.
    let directoryURL: URL

    /// The loaded `course_config.json` for this course.
    let configuration: CourseConfiguration

    // MARK: - Computed properties

    var id: String {
        return code
    }

    /// The timetable section numbers for this course, from the configuration.
    var sectionNumbers: [Int] {
        return configuration.sectionNumbers
    }

    var configFileURL: URL {
        return directoryURL.appendingPathComponent("course_config.json")
    }

    // MARK: - Initializer

    init(code: String, directoryURL: URL, configuration: CourseConfiguration) {
        self.code = code
        self.directoryURL = directoryURL
        self.configuration = configuration
    }
}

File: windows-app/Plantoir.Core/Models/Course.cs
namespace Plantoir.Core.Models;

/// <summary>One course/club folder under &lt;workspace&gt;/courses/, plus its loaded config.</summary>
public sealed class Course
{
    /// <summary>The course code — also the folder name (e.g. ICS3U).</summary>
    public string Code { get; }

    /// <summary>&lt;workspace&gt;/courses/&lt;CODE&gt;</summary>
    public string DirectoryPath { get; }

    public CourseConfiguration Configuration { get; }

    public Course(string code, string directoryPath, CourseConfiguration configuration)
    {
        Code = code;
        DirectoryPath = directoryPath;
        Configuration = configuration;
    }

    public List<int> SectionNumbers => Configuration.SectionNumbers;

    public string ConfigFilePath => Path.Combine(DirectoryPath, "course_config.json");

    public string SectionDirectory(int number) => Path.Combine(DirectoryPath, "section" + number);
}
